### PR TITLE
Dawn 588 update mongodb gh#1403

### LINKED ISSUE #1403: DAWN-588 ⁃ Update mongodb database plugin for Dawn 3.0
The block structure, transaction structure, native contracts, etc. have changed in Dawn 3.0 which impacts the mongodb plugin and therefore the web wallet.

Update the mongodb plugin to continue to have the information needed by the web wallet.

ATC:

* Run tests/eosd_run_test.py --mongodb

### PR DESCRIPTION
**Pull request by heifner**
* Updated mongodb plugin to process reversible and irreversible blocks so that deferred and inline actions can be captured.
* Added action_trace to mongodb
* Added deferred transaction attributes to mongodb: sender_id, sender, execute_after
* Removed the explicit list of transactions from block. Available via a query.
* Removed the explicit list of actions from transaction. Available via a query.

Resolves #1403

### CODE REVIEW

File: plugins/mongo_db_plugin/mongo_db_plugin.cpp [hunk 1]
```diff
@@ -123,31 +127,69 @@ void mongo_db_plugin_impl::applied_irreversible_block(const signed_block& block)
...
+   } catch (fc::exception& e) {
+      elog("FC Exception while applied_block ${e}", ("e", e.to_string()));
+   } catch (std::exception& e) {
+      elog("STD Exception while applied_block ${e}", ("e", e.what()));
+   } catch (...) {
+      elog("Unknown exception while applied_block");
+   }
+}
+
 void mongo_db_plugin_impl::consum_blocks() {
```
Comment (wanderingbort) [inline plugins/mongo_db_plugin/mongo_db_plugin.cpp]: consum**e** ?

File: plugins/mongo_db_plugin/mongo_db_plugin.cpp [hunk 2]
```diff
@@ -379,6 +544,59 @@ void mongo_db_plugin_impl::_process_irreversible_block(const signed_block& block
...
+   auto ir_block = find_block(blocks, block_id_str);
+
+   document update_block{};
+   update_block << "$set" << open_document << "pending" << b_bool{false}
+               << "updatedAt" << b_date{now}
+               << close_document;
+
+   blocks.update_one(document{} << "_id" << ir_block.view()["_id"].get_oid() << finalize, update_block.view());
+
+   for (const auto& packed_trx : block.input_transactions) {
```
Comment (wanderingbort) [inline plugins/mongo_db_plugin/mongo_db_plugin.cpp]: this seems to only mark the user generated transactions as irreversible.

When will deferred transactions be marked irreversible?  Should this go through the region/cycle/shard summaries on the irreversible block to derive transaction id's to mark irreversible?

File: plugins/mongo_db_plugin/mongo_db_plugin.cpp (hunk 1 above)
Comment (heifner) [reply plugins/mongo_db_plugin/mongo_db_plugin.cpp]: Yep. And I misspelled it throughout.

**Comment by heifner** [comment]
Updated with requested changes and fixed merge issue.

**Pull request merged**